### PR TITLE
Adopt the Extended Heat Index calculation in zone resilience

### LINKED ISSUE #10538: Current HI doesn't cover extreme conditions, propose to improve with extended heat index
Issue overview
--------------
The heat index reflects human perceived hotness considering both temperature and humidity. The current heat index calculation in EnergyPlus is based on Steadman’s method, which gives nonphysical results in extreme weather conditions. The [extended heat index](https://journals.ametsoc.org/view/journals/apme/61/10/JAMC-D-22-0021.1.xml) fixes this issue and outputs a more realistic heat index. It can also be associated with the health outcomes (heat stroke, etc.) of an idealized human, helpful in the quantification of damages from extreme weather.

The implementation is based on code on [this page](https://romps.berkeley.edu/papers/pubs-2020-heatindex.html)

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by yujiex**
Pull request overview
---------------------
 - Fixes #10538
 - This PR implements the extended HI. The code is adapted from python code on [this page](https://romps.berkeley.edu/papers/pubs-2020-heatindex.html)

### Introduction

The heat index reflects human perceived hotness considering both temperature and humidity. The current heat index calculation in EnergyPlus is based on Steadman’s method, which gives nonphysical results in extreme weather conditions. The [extended heat index](https://journals.ametsoc.org/view/journals/apme/61/10/JAMC-D-22-0021.1.xml) fixes this issue and outputs a more realistic heat index. It can also be associated with the health outcomes (heat stroke, etc.) of an idealized human, helpful in the quantification of damages from extreme weather.

The following plot shows the absolute and percent difference between the extended HI implemented in this feature branch and the HI in the develop branch

absolute_diff_newEP_old

percent_diff_newEP_old

The difference are mainly in regions with higher temperature and humidity. See the following heat index plot from the [National Weather Service Website](https://www.weather.gov/ama/heatindex)

image

The following is the extended heat index implemented in this feature branch.

extendedHI_EP

There are also some slight differences between the extended HI implemented in this feature and the heatindex.py results by Lu and Romps. The maximum difference is around e-5. This difference is originated from the root solver difference between EnergyPlus and heatindex.py

absolute_diff_newEP_new

percent_diff_newEP_new

### Regression diffs

The following files have regression diffs. The ESO diffs all happen in the "ZONE ONE:Zone Heat Index [C]" variable. The table diffs are in heat index hours. Note that the heat index hours counts the number of hours with heat index in a certain range. These changes are expected as this feature changes the heat index calculation method.
 
image

Using "1ZoneUncontrolled_Win_ASH55_Thermal_Comfort.idf" as an example, the diff looks like the following

image

RH is 100 or close to 100. When temperature is in the range of 90F to 100F, the HI computed using the Steadman and the extended heat index method can differ by 0 to 12C. 

Some other diffs are in resilience report summary tables. These tables are produced when "AllSummary" or "ThermalResilienceSummary" are present in "Output:Table:SummaryReports". For example, this is the diff of test file "_5ZoneAirCooled_LeapYear_annual.idf". The diffs are due to the change in the heat index calculation in this feature branch.
image

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by yujiex** [comment]
Hi @Myoldmopar and @mjwitte this is ready for review. Thanks.

File: doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex
Comment (Myoldmopar) [inline doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex]: There is a LaTeX error here with the & character, it needs to be escaped.

File: src/EnergyPlus/CMakeLists.txt
Comment (Myoldmopar) [inline src/EnergyPlus/CMakeLists.txt]: I don't like the lower cased names...  I mean in general I am fine with it, but it looks super weird against all the other PascalCased names.  I'm going to change this if I push changes up.

File: src/EnergyPlus/extendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/extendedHI.cc]: Are these supposed to be debug messages that can be removed?  I dislike seeing std::cout directly in E+ code, even if it's only meant to be used during debugging.

File: src/EnergyPlus/extendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/extendedHI.cc]: Are you asking a function to return a string, and then doing a string comparison to decide which code path to follow?  For one thing, this arrangement seems way way overkill just to calculate heat index.  And second, it cannot be a string, you need to _at least_ use an enum.  But I really feel this whole thing could be way simpler.

File: src/EnergyPlus/extendedHI.hh
Comment (Myoldmopar) [inline src/EnergyPlus/extendedHI.hh]: No newline here.

File: tst/EnergyPlus/unit/extendedHI.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/extendedHI.unit.cc]: I'm inclined to say all the values in this unit test are _waaay_ too precise.  We cannot know these values so precisely, so to test against them is futile.  I see you have tolerance set to e-8, which is pretty tight here.  But there are other ones where tolerance is more like e-4, yet the values you compare against have like 12+ decimals.  It's not the end of the world, it's just a lot of numbers here :)

File: tst/EnergyPlus/unit/extendedHI.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/extendedHI.unit.cc]: Fix me

File: src/EnergyPlus/extendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/extendedHI.cc]: Ummmm making a copy of the whole root solver object each time we calculate heat index.... that's a no.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
OK, so my main concern is around the complexity of the heat index calculation.  Asking functions to return pairs and tuples that include strings, and then checking the string to do different operations, is all .... a lot.  I'm also highly concerned about making a copy of the entire root solver object.  This needs some work.  Simplification work.

I have less critical other concerns: a LaTeX error, the filename should ideally be ExtendedHI.* so that it matches our existing pattern, the unit test precision gives some feels but might be fine.

File: doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex
Comment (yujiex) [reply doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex]: Thanks. I just escaped it.

File: src/EnergyPlus/CMakeLists.txt
Comment (yujiex) [reply src/EnergyPlus/CMakeLists.txt]: Thanks. I just changed the filenames to PascalCased names.

File: src/EnergyPlus/extendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/extendedHI.cc]: They are debugging messages. I will remove them.

File: src/EnergyPlus/extendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/extendedHI.cc]: I will simplify this

File: src/EnergyPlus/extendedHI.hh
Comment (yujiex) [reply src/EnergyPlus/extendedHI.hh]: Thanks, I just added newline at the end.

File: tst/EnergyPlus/unit/extendedHI.unit.cc
Comment (yujiex) [reply tst/EnergyPlus/unit/extendedHI.unit.cc]: I will unify the tolerance and the values here.

These are output from this python file. I was trying to ideally match their result. 
[heatindex.py.zip](https://github.com/user-attachments/files/16146813/heatindex.py.zip)

The e-4 part is due to the difference in the root solvers between EnergyPlus and the source python file  (when EnergyPlus uses the bisection one as well).
The following is a illustration of the difference in a somewhat extreme case
image
Here is the difference in the code
image

File: src/EnergyPlus/extendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/extendedHI.cc]: I see. I changed it to this

```
auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverAlgorithm::Bisection;
// do stuff
state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverBackup;
```

**Comment by yujiex** [comment]
Hi @Myoldmopar, I've simplified the functions some more and addressed other comments as well. Please see if they look okay. Thanks.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc]: This is still making a copy of the root solver each call to heatIndex.  Which happens for each zone at each time step.  This is not OK.  I get the feeling you shouldn't use the HVACSystemRootSolver for this.  Instead just use SolveRoot?  

As an aside, I still think the find_eqvar with tuple and returning different things is still a bit too complex just to calculate heat index.  At this point I wouldn't hold it up just for that, but if it could be simplified further, that would be great.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
This is still making a copy of the root solver object in each call to heat index, and that needs to be addressed before merging.  If you are needing to simply find a zero, you can do this which is what we do in a ton of places:

```
int constexpr maxIter = 1000;
Real64 constexpr xMin = -10.0;
Real64 constexpr xMax = 10.0;
Real64 constexpr accuracy = 0.001;
Real64 someTargetCondition = abc;
auto f = [&state] (Real64 x) { return x - someTargetCondition; };
int flag = 0;
Real64 x = 0.0;
General::SolveRoot(state, accuracy, maxIter, flag, x, f, xMin, xMax);
```

Is it possible to use that here rather than the HVACSolver stuff?

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc]: Hi Edwin, I was using `General::SolveRoot` to solve the roots. The purpose of the part involving `HVACSystemRootSolver` is just so that the root searching order inside the `General::SolveRoot` function follows the bisection method, which is what was used in the original python implementation. Maybe I'll just remove this part and hopefully they don't differ much when using different root finding schemes. So far the tests in the unit test file still passes if this part is removed.

I will try to simplify `find_eqvar` some more

Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc]: @Myoldmopar I just changed `find_eqvar` to two functions each returning a single value.

Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc]: btw I also just printed out the size of the things involved here. They're all 4 byte.
```
auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
```
image

Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc]: Removing the 
```
        auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
        auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
        state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverAlgorithm::Bisection;

...
        state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverBackup;
```
seems to cause some NaN in the unit test `HeatBalanceSurfaceManager_TestResilienceMetricReport`, when calculating this: `ExtendedHI::heatindex(state, 298.15, 0.5) `. Also in unit test `HeatBalanceSurfaceManager_TestThermalResilienceReportRepPeriod`

**Comment by Myoldmopar** [comment]
@yujiex I apologize for sending you down a distraction path here.  When I saw that you were making a copy of 
`state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver` I immediately thought it was making a copy of a root solver.  Like a whole root solver.  Since that's the name of the variable :) 

When you said it was only 4 bytes, that of course made things different.  It's just a freaking enum.  Sorry for distracting you with this, please revert any unnecessary changes and let's get this merged.

**Comment by yujiex** [comment]
> @yujiex I apologize for sending you down a distraction path here. When I saw that you were making a copy of `state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver` I immediately thought it was making a copy of a root solver. Like a whole root solver. Since that's the name of the variable :)
> 
> When you said it was only 4 bytes, that of course made things different. It's just a freaking enum. Sorry for distracting you with this, please revert any unnecessary changes and let's get this merged.

It's totally fine. It might be more clear if the variable was named something like RootSearchingMode. Currently it sounds like very memory intensive.

I just reverted the changes.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Pulling develop in here as well to get fresh CI results.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc:110]: This should be constexpr-able right?  And I will assume for now that these are indeed used in multiple locations where it makes sense to keep them out here in the namespace.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc:140]: In places where you can't `constexpr`, it's still good to `const`.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc:177]: Yes, this is the way.  If you want to make it slightly clearer, you _can_ define the lambda first so that you have plenty of room to name things nicely, and then just pass the lambda in as the argument.  It's fine as an argument here though.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc:236]: This is definitely getting pretty long to just be a lambda argument.  I'd consider making this a nicely named function with clear arguments, etc., and then just have the lambda call that function.  Debugging an issue with the arithmetic in this line would be quite difficult.

File: src/EnergyPlus/ExtendedHI.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.cc:534]: In general, this is so much better than it was, great job there.  I am still quite surprised to find the sheer amount of SolveRoot calls inside this module which is just trying to calculate heat index.  Anyway, thanks for the cleanups.

File: src/EnergyPlus/ExtendedHI.hh
Comment (Myoldmopar) [inline src/EnergyPlus/ExtendedHI.hh:70]: I'm not convinced Invalid and Num are needed here.  This enum is not paired with any input strings, right?  I would suggest removing Invalid and Num, moving this enum class into the cc file (since it's not used here in the header), and then adding this enum to the exclude list [here](https://github.com/NREL/EnergyPlus/blob/d6e49668d138eea326cc7917b75215fbca7ecf64/scripts/dev/check_for_malformed_enums.py#L109) and [here](https://github.com/NREL/EnergyPlus/blob/d6e49668d138eea326cc7917b75215fbca7ecf64/scripts/dev/check_for_malformed_enums.py#L132)

File: tst/EnergyPlus/unit/ExtendedHI.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ExtendedHI.unit.cc:370]: Interesting.  I guess this is fine.  I start getting nervous when the amount of logic inside the unit tests grows.  Like now I need to test this function as well...  But I get the intent here, so I'll just move on.

File: tst/EnergyPlus/unit/ExtendedHI.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ExtendedHI.unit.cc:376]: I feel like we definitely have this somewhere else so you don't need your own.

File: tst/EnergyPlus/unit/ExtendedHI.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ExtendedHI.unit.cc:607]: So where did these values come from?  I am guessing they are just generated from running the code, not from an external source.  If so, this is OK, but not great.  If these values are from a separately validated source, then that is a nice verification.

**Review by Myoldmopar — APPROVED** [review]
I'm not requesting any specific changes.  I left some comments that you can respond to, but I think this is probably fine to go in.

**Comment by yujiex** [comment]
> I'm not requesting any specific changes. I left some comments that you can respond to, but I think this is probably fine to go in.

Thanks, @Myoldmopar. I will address these today

**Comment by Myoldmopar** [comment]
I pushed the branch with develop pulled in and will wait for CI to give fresh results.  Once that's done, this can merge.  @yujiex if you want to address any of my comments via code changes, they can come later.  For now just respond to them as comments however much you want.

**Comment by yujiex** [comment]
> I pushed the branch with develop pulled in and will wait for CI to give fresh results. Once that's done, this can merge. @yujiex if you want to address any of my comments via code changes, they can come later. For now just respond to them as comments however much you want.

I see. Thanks @Myoldmopar

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc:110]: yes. It should be. Yeah, some stuff like `Pc` is used in a few functions.

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc:140]: I see. Good point

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc:177]: Make sense. Probably better as their own function in front

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc:236]: Yeah, indeed. This is probably better separated out.

File: src/EnergyPlus/ExtendedHI.cc
Comment (yujiex) [reply src/EnergyPlus/ExtendedHI.cc:534]: Thanks for reviewing. Glad I've been able to clean it out a bit. It does seem a pretty involved calculation process in their paper as well.

**Comment by Myoldmopar** [comment]
Still happy here locally with develop pulled in, merging.  Thanks @yujiex

**Comment by yujiex** [comment]
Thanks @Myoldmopar !

**Pull request merged**
